### PR TITLE
Add linter for DOI type citation.

### PR DESCRIPTION
**Pull request by mvdbeek**
I guess this is useful on its own, and may prove more valuable if we support plugging tool information from http://bio.tools into `planemo tool_init`.

### CODE REVIEW

**Comment by jmchilton** [comment]
Very nice - I love this!

**Comment by mvdbeek** [comment]
Thanks @jmchilton - seems like I created a [false positive](https://travis-ci.org/galaxyproject/planemo/jobs/131742040#L673) in the tests though

File: tests/data/tools/fail_doi.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<tool id="copy" name="Copy Dataset" version="1.0">
+    <description>copies a dataset</description>
```
Comment (jmchilton) [inline tests/data/tools/fail_doi.xml]: The test framework just ensures all ``fail_`` tools fail linting - just rename to invaild_doi or something and the failing test will work.

File: planemo/lint.py [hunk 2]
```diff
@@ -7,8 +8,43 @@
...
+        is_doi(publication, lint_ctx)
+
+
+def find_dois_for_xml(root):
+    dois = []
+    for element in root.root.findall("citations"):
+        for citation in list(element):
+            if "type" in citation.attrib:
+                if citation.attrib["type"] == "doi":
+                    dois.append(citation.text)
```
Comment (nsoranzo) [inline planemo/lint.py]: I'd substitute the previous 3 lines with:

```python
            if citation.tag == 'citation' and citation.attrib.get('type', '') == 'doi':
                dois.append(citation.text)
```

File: planemo/lint.py [hunk 3]
```diff
@@ -7,8 +8,43 @@
...
+    r = requests.get(url)
+    if r.status_code == 200:
+        if publication_id != doiless_publication_id:
+            lint_ctx.error("%s is valid, but galaxy expects DOI without 'doi:' prefix" % publication_id)
+        else:
+            lint_ctx.info("%s is a valid DOI" % publication_id)
+    elif r.status_code == 404:
+        lint_ctx.error("%s is not a valid DOI" % publication_id)
+    else:
+        lint_ctx.info("dx.doi returned unexpected status code %d" % r.status_code)
```
Comment (nsoranzo) [inline planemo/lint.py]: `lint_ctx.warn` ?

File: planemo/lint.py [hunk 4]
```diff
@@ -7,8 +8,43 @@
...
+    """
+    Check if dx.doi knows about the publication_id
+    """
+    base_url = "http://dx.doi.org"
+    doiless_publication_id = publication_id.split("doi:", 1)[-1]
+    url = "%s/%s" % (base_url, doiless_publication_id)
+    r = requests.get(url)
+    if r.status_code == 200:
+        if publication_id != doiless_publication_id:
+            lint_ctx.error("%s is valid, but galaxy expects DOI without 'doi:' prefix" % publication_id)
```
Comment (nsoranzo) [inline planemo/lint.py]: s/galaxy/Galaxy/

File: tests/data/tools/fail_doi.xml (hunk 1 above)
Comment (mvdbeek) [inline tests/data/tools/fail_doi.xml]: Ahh, indeed, that did the trick, thanks!

File: tests/test_lint.py [hunk 5]
```diff
@@ -64,3 +64,8 @@ def test_empty_cdata(self):
         empty_cdata = os.path.join(TEST_TOOLS_DIR, "empty_cdata.xml")
         lint_cmd = ["lint", "--skip", "citations,help", empty_cdata]
         self._check_exit_code(lint_cmd, exit_code=0)
+
+    def test_lint_doi(self):
+        fail_doi = os.path.join(TEST_TOOLS_DIR, "invalid.xml_doi.xml")
```
Comment (nsoranzo) [inline tests/test_lint.py]: s/invalid.xml_doi.xml/invalid_doi.xml/ ?

**Comment by jmchilton** [comment]
Awesome thanks!

File: planemo/lint.py [hunk 6]
```diff
@@ -7,8 +8,42 @@
...
+    """
+    Check if dx.doi knows about the publication_id
+    """
+    base_url = "http://dx.doi.org"
+    doiless_publication_id = publication_id.split("doi:", 1)[-1]
+    url = "%s/%s" % (base_url, doiless_publication_id)
+    r = requests.get(url)
+    if r.status_code == 200:
+        if publication_id != doiless_publication_id:
+            lint_ctx.error("%s is valid, but Galaxy expects DOI without 'doi:' prefix" % publication_id)
```
Comment (nsoranzo) [inline planemo/lint.py]: Are you sure about this? @bgruening pointed out in https://github.com/galaxyproject/tools-iuc/pull/768#issuecomment-221633995 that the initial 'doi:' seems to work fine.

Comment (mvdbeek) [inline planemo/lint.py]: I just tested it again, the doi breaks the display of the reference with my terribly out-of-date planemo instance (September 2015 I believe), but newer galaxy can deal with the `doi:`.  I guess we could tone this down to an info?

Comment (nsoranzo) [inline planemo/lint.py]: Something like:
```python
lint_ctx.warn("%s is valid, but old Galaxy versions expected DOI without 'doi:' prefix" % publication_id)
```

?

Comment (mvdbeek) [inline planemo/lint.py]: Yes ... but given those old instances will disappear over time, maybe we can remove the warning altogether? Alternatively we should at least figure out how old is `old` :/

Comment (nsoranzo) [inline planemo/lint.py]: Up to you or @jmchilton, I don't have a strong opinion on this.

Comment (jmchilton) [inline planemo/lint.py]: I have no clue what changed to enable those - I just rewalked through the code and I can't even figure out why it works at all with the leading doi: - I don't think it should. Anyway - I definitely consider a leading ``doi:`` a tool bug and on a purely stylistic point somewhat redundant given the citation type is "doi". I think ``warn`` is entirely appropriate here.

**Pull request merged**